linalg/hypre.cpp

```diff
@@ -1532,15 +1532,27 @@ HypreParMatrix *HypreParMatrix::ExtractSubmatrix(const Array<int> &indices,
    int local_num_vars = hypre_CSRMatrixNumRows(hypre_ParCSRMatrixDiag(A));
 
    // Form hypre CF-splitting array designating submatrix as F-points (-1)
+#ifdef hypre_IntArrayData
+   hypre_IntArray *CF_marker;
+
+   CF_marker = hypre_IntArrayCreate(local_num_vars);
+   hypre_IntArrayInitialize(CF_marker);
+   hypre_IntArraySetConstantValues(CF_marker, 1);
+#else
    Array<HYPRE_Int> CF_marker(local_num_vars);
    CF_marker = 1;
+#endif
    for (int j=0; j<indices.Size(); j++)
    {
       if (indices[j] > local_num_vars)
       {
          MFEM_WARNING("WARNING : " << indices[j] << " > " << local_num_vars);
       }
+#ifdef hypre_IntArrayData
+      hypre_IntArrayData(CF_marker)[indices[j]] = -1;
```

Comment:
If I understand the `hypre_IntArray` correctly, when HYPRE is built with CUDA, the pointer `hypre_IntArrayData(CF_marker)` is a CUDA pointer, right? If so, we can't use it here on CPU.

Reply:
@v-dobrev This depends on whether hypre is built using unified memory. Currently, there are still many features that require unified memory (see my comment below regarding `hypre_ParCSRMatrixExtractSubmatrixFC`), so I think it's fine to leave this code as is for now and require the use of unified memory when using cuda.

Reply:
If that is the case, it is best for this method to error out if HYPRE is built with CUDA without UVM:
```
#if defined(HYPRE_USING_CUDA) && !defined(HYPRE_USING_UNIFIED_MEMORY)
   MFEM_ABORT("This method requires HYPRE built with UVM when HYPRE is using CUDA!");
#endif
```
I now see that the real reason we need to use `hypre_IntArray *` is that `hypre_BoomerAMGCoarseParms` now expects `CF_marker` to be of that type instead of `HYPRE_Int *`.

Reply:
@v-dobrev That's correct. And I think your error handling suggestion is also a good idea.

Reply:
I just added the above check.

Reply:
Looking at this more closely, the CF_marker created here is ONLY used for the call to `hypre_ParCSRMatrixExtractSubmatrixFC`, so there is no reason for this to be allocated on the device since that function is only CPU code right now. So the CF_marker should be initialized on the host:
`hypre_IntArrayInitialize_v2( CF_marker, HYPRE_MEMORY_HOST )`
This will remove the need for the check above (i.e. unified memory is not required here). Sorry for the initial confusion, I just didn't see the full context for the usage here.

Reply:
What about the use of `CF_marker` in `hypre_BoomerAMGCoarseParms`?

Reply:
That function will work regardless of the memory location of CF_marker.

Reply:
Okay, I updated the method to work on host, in particular to allocate `CF_marker` always on host.